src/pathpyG/core/MultiOrderModel.py

```diff
@@ -278,3 +282,280 @@ def from_DAGs(
+        _, counts = torch.unique(dag_graph.node_sequence, return_counts=True)
+        node_emission_probabilities = counts / counts.sum()
+        return torch.mul(frequencies, torch.log(node_emission_probabilities[start_ixs])).sum().item()
+
+    def get_intermediate_order_log_likelihood(self, dag_graph: DataBatch, order: int) -> float:
+        """
+        Compute the intermediate order log likelihood.
+
+        Args:
+            m (MultiOrderModel): Multi-order model.
+            dag_graph (DataBatch): Input DAG graph data.
+            order (int): Order of the intermediate log likelihood.
+
+        Returns:
+            float: Intermediate order log likelihood.
+        """
+        # Get frequencies
+        # TODO: put this tensor directly in dag_graph (intead of edge_weight) and remove the following line
+        # getting the index of the last edge of each path (to be used to extract weights)
+        last_edge_ixs = dag_graph.ptr[1:] - torch.arange(2,dag_graph.ptr.shape[0]+1)
+        frequencies = dag_graph.edge_weight[last_edge_ixs]
+
+        # Get intermediate HO nodes ixs
+        mask = torch.ones(dag_graph.num_nodes, dtype=bool)
+        mask[dag_graph.edge_index[1]] = False
+        ixs = torch.where(mask)[0]
+        num_ixs = ixs.shape[0]
+        ho_intermediate_ixs = ixs - torch.arange(num_ixs) * order
+
+        return (
+            torch.mul(
+                frequencies,
+                torch.log(
+                    compute_transition_probabilities(self.layers[order])[
+                        self.layers[order + 1].data.inverse_idx[ho_intermediate_ixs]
+                    ]
+                ),
+            )
+            .sum()
+            .item()
+        )
+
+    def get_mon_log_likelihood(self, dag_graph: DataBatch, max_order: int = 1) -> float:
+        """
+        Compute the likelihood of the walks given a multi-order model.
+
+        Args:
+            m (MultiOrderModel): The multi-order model.
+            dag_graph (DataBatch): Dataset containing the walks.
+            max_order (int, optional): The maximum order up to which model layers
+                shall be taken into account. Defaults to 1.
+
+        Returns:
+            float: The log likelihood of the walks given the multi-order model.
+        """
+        llh = 0
+
+        # Adding likelihood of zeroth order
+        llh += self.get_zeroth_order_log_likelihood(dag_graph)
+
+        # Adding the likelihood for all the intermediate orders
+        for order in range(1, max_order):
+            llh += self.get_intermediate_order_log_likelihood(dag_graph, order)
+
+        # Adding the likelihood of highest/stationary order
+        if max_order > 0:
+            llh += (
+                (
+                    torch.log(compute_transition_probabilities(self.layers[max_order]))
+                    * self.layers[max_order].data.edge_weight
+                )
+                .sum()
+                .item()
+            )
```

Comment:
Again, some intermediate steps would be nice for readability.

Reply:
Or add the formula as comment as done below

Reply:
added intermediate steps